axlearn/common/trainer_config_modifier.py

```diff
@@ -146,6 +186,111 @@ def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
         return cfg
 
 
+class ModelConfigModifier(ConfigModifier):
+    """Update the model config for the trainer config."""
+
+    @config_class
+    class Config(ConfigModifier.Config):
+        """Configure ModelConfigModifier.
+
+        Attributes:
+            model_cfg_modifications: A mapping from module path
+                (e.g. `model.decoder.transformer.layer`) to a Config.
+        """
+
+        model_cfg_modifications: Required[Dict[str, ConfigBase]] = REQUIRED
+
+    def __init__(self, cfg: Config):
+        super().__init__(cfg)
+        cfg = self.config
+        self._model_cfg_modifications = cfg.model_cfg_modifications
+
+    def _merge_configs(self, target_cfg: ConfigBase, found_module: ConfigBase) -> ConfigBase:
+        """Merge configurations from the config being replaced on a best effort basis.
+        Args:
+            target_cfg: configuration that will replace found_module.
+            found_module: existing configuration whose class will be replaced
+                but it's confguration will be merged with target_cfg.
+
+        Returns:
+            The modified config.
+
+        """
+        for key in target_cfg.keys():
+            if key == "klass":
+                continue
+            elif hasattr(found_module.module, key) and hasattr(target_cfg, key):
+                setattr(target_cfg, key, getattr(found_module.module, key))
+        return target_cfg
+
+    def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
+        """Overwrite the model config of the specified modules.
+
+        Args:
+            cfg: The trainer config to be modified.
+
+        Raises:
+            ValueError: The target module is not found.
+
+        Returns:
+            The modified trainer config.
+        """
+
+        for module_name, model_cfg in self._model_cfg_modifications.items():
+            found_module = _find_target_module(module_name, cfg)
```

Comment:
In utils.py we have `get_recursively` and `set_recursively` for `Nested[...]`. I wonder if it will be useful to add corresponding methods to ConfigBase. Then we can do something like:

```suggestion
        for cfg_path, cfg_modification in self._model_cfg_modifications.items():
            child_cfg = cfg.get_recursively(cfg_path)
            child_cfg = cfg_modification(child_cfg, path=cfg_path)
            cfg.set_recursively(cfg_path, value=child_cfg)
```

Reply:
I wonder if an alternative (which aims to simplify the `ConfigBase` api) is to do something similar to Python's `sorted`; we allow `utils.get_resursively` to take a value fn:
```
# Default behavior is to use key lookup:
utils.get_recursively(..., value_fn=lambda k,v: v[k])

# Custom behavior can be attribute lookup:
utils.get_recursively(..., value_fn=lambda k,v: getattr(v,k))
```
A benefit is that other non-config instances can also leverage `get_recursively`.

Reply:
Hi @markblee , maybe we can do this in a follow-up PR?